File: python-modular/examples/all_classifier.py
#!/usr/bin/env python
"""
Explicit examples on how to use the different classifiers
"""

from numpy import double, array, floor, concatenate
from numpy.random import rand, seed, permutation
from shogun.Kernel import GaussianKernel, WeightedDegreeStringKernel
from shogun.Distance import EuclidianDistance
from shogun.Features import *
from shogun.Classifier import *

def get_clouds (num, num_feats, num_vec):
	data=[rand(num_feats, num_vec)+x/2 for x in xrange(num)]
	cloud=concatenate(data, axis=1)
	return array([permutation(x) for x in cloud])

def get_dna ():
	acgt=array(['A', 'C', 'G','T'])
	len_acgt=len(acgt)
	rand_train=[]
	rand_test=[]

	for i in xrange(11):
		str1=[]
		str2=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
			str2.append(acgt[floor(len_acgt*rand())])
		rand_train.append(''.join(str1))
	rand_test.append(''.join(str2))
	
	for i in xrange(6):
		str1=[]
		for j in range(60):
			str1.append(acgt[floor(len_acgt*rand())])
	rand_test.append(''.join(str1))

	return {'train': rand_train, 'test': rand_test}

###########################################################################
# kernel-based SVMs
###########################################################################

def svm_light ():
	print 'SVMLight'

	data=get_dna()
	feats_train=StringCharFeatures(DNA)
	feats_train.set_string_features(data['train'])
	feats_test=StringCharFeatures(DNA)
	feats_test.set_string_features(data['test'])
	degree=20

	kernel=WeightedDegreeStringKernel(feats_train, feats_train, degree)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=SVMLight(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def libsvm ():
	print 'LibSVM'

	num_feats=9
	data=get_clouds(2, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=2
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=LibSVM(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def gpbtsvm ():
	print 'GPBTSVM'

	num_feats=9
	data=get_clouds(2, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=8
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=GPBTSVM(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def mpdsvm ():
	print 'MPDSVM'

	num_feats=9
	data=get_clouds(2, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=MPDSVM(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def libsvm_multiclass ():
	print 'LibSVMMultiClass'

	num_feats=9
	data=get_clouds(5, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(5, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=8
	lab=[double(x) for x in xrange(feats_train.get_num_vectors())]
	labels=Labels(array(lab))

	svm=LibSVMMultiClass(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def libsvm_oneclass ():
	print 'LibSVMOneClass'

	num_feats=9
	data=get_clouds(2, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=1

	svm=LibSVMOneClass(C, kernel)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

def gmnpsvm ():
	print 'GMNPSVM'

	num_feats=9
	data=get_clouds(4, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(4, num_feats, 17)
	feats_test=RealFeatures(data)
	width=2.1
	kernel=GaussianKernel(feats_train, feats_train, width)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=1
	lab=[double(x) for x in xrange(feats_train.get_num_vectors())]
	labels=Labels(array(lab))

	svm=GMNPSVM(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)
	svm.classify().get_labels()

###########################################################################
# run with batch or linadd on LibSVM
###########################################################################

def do_batch_linadd ():
	print 'LibSVM batch'

	data=get_dna()
	feats_train=StringCharFeatures(DNA)
	feats_train.set_string_features(data['train'])
	feats_test=StringCharFeatures(DNA)
	feats_test.set_string_features(data['test'])
	degree=20

	kernel=WeightedDegreeStringKernel(feats_train, feats_train, degree)

	C=0.017
	epsilon=1e-5
	tube_epsilon=1e-2
	num_threads=2
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=LibSVM(C, kernel, labels)
	svm.set_epsilon(epsilon)
	svm.set_tube_epsilon(tube_epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.train()

	kernel.init(feats_train, feats_test)

	print 'LibSVM Objective: %f num_sv: %d' % \
		(svm.get_objective(), svm.get_num_support_vectors())
	svm.set_batch_computation_enabled(False)
	svm.set_linadd_enabled(False)
	svm.classify().get_labels()

	svm.set_batch_computation_enabled(True)
	svm.classify().get_labels()

###########################################################################
# linear SVMs
###########################################################################

def subgradient_svm ():
	print 'SubGradientSVM'

	num_feats=11
	data=get_clouds(2, num_feats, 11)
	realfeat=RealFeatures(data)
	feats_train=SparseRealFeatures()
	feats_train.obtain_from_simple(realfeat)
	data=get_clouds(2, num_feats, 21)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)

	C=0.42
	epsilon=1e-3
	num_threads=1
	max_train_time=1.
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=SubGradientSVM(C, feats_train, labels)
	svm.set_epsilon(epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.set_bias_enabled(False)
	svm.set_max_train_time(max_train_time)
	svm.train()

	svm.set_features(feats_test)
	svm.classify().get_labels()

def svmocas ():
	print 'SVMOcas'

	num_feats=11
	data=get_clouds(2, num_feats, 12)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)
	data=get_clouds(2, num_feats, 21)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)

	C=0.42
	epsilon=1e-5
	num_threads=1
	lab=rand(feats_test.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=SVMOcas(C, feats_test, labels)
	svm.set_epsilon(epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.set_bias_enabled(False)
	svm.train()

	svm.set_features(feats_test)
	svm.classify().get_labels()

def liblinear ():
	print 'LibLinear'

	num_feats=11
	data=get_clouds(2, num_feats, 10)
	realfeat=RealFeatures(data)
	feats_train=SparseRealFeatures()
	feats_train.obtain_from_simple(realfeat)
	data=get_clouds(2, num_feats, 25)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)

	C=0.42
	epsilon=1e-5
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=LibLinear(C, feats_train, labels)
	svm.set_epsilon(epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.set_bias_enabled(True)
	svm.train()

	svm.set_features(feats_test)
	svm.classify().get_labels()

def svmlin ():
	print 'SVMLin'

	num_feats=11
	data=get_clouds(2, num_feats, 11)
	realfeat=RealFeatures(data)
	feats_train=SparseRealFeatures()
	feats_train.obtain_from_simple(realfeat)
	data=get_clouds(2, num_feats, 25)
	realfeat=RealFeatures(data)
	feats_test=SparseRealFeatures()
	feats_test.obtain_from_simple(realfeat)

	C=0.42
	epsilon=1e-5
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	svm=SVMLin(C, feats_train, labels)
	svm.set_epsilon(epsilon)
	svm.parallel.set_num_threads(num_threads)
	svm.set_bias_enabled(True)
	svm.train()

	svm.set_features(feats_test)
	svm.get_bias()
	svm.get_w()
	svm.classify().get_labels()

###########################################################################
# misc classifiers
###########################################################################

def perceptron ():
	print 'Perceptron'

	num_feats=9
	data=get_clouds(2, num_feats, 11)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 17)
	feats_test=RealFeatures(data)

	learn_rate=1.
	max_iter=1000
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	perceptron=Perceptron(feats_train, labels)
	perceptron.set_learn_rate(learn_rate)
	perceptron.set_max_iter(max_iter)
	perceptron.parallel.set_num_threads(num_threads)
	perceptron.train()

	perceptron.set_features(feats_test)
	perceptron.classify().get_labels()

def knn ():
	print 'KNN'

	num_feats=9
	data=get_clouds(2, num_feats, 10)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 18)
	feats_test=RealFeatures(data)
	distance=EuclidianDistance()

	k=3
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	knn=KNN(k, distance, labels)
	knn.parallel.set_num_threads(num_threads)
	knn.train()

	distance.init(feats_train, feats_test)
	knn.classify().get_labels()

def lda ():
	print 'LDA'

	num_feats=9
	data=get_clouds(2, num_feats, 14)
	feats_train=RealFeatures(data)
	data=get_clouds(2, num_feats, 27)
	feats_test=RealFeatures(data)

	gamma=3
	num_threads=1
	lab=rand(feats_train.get_num_vectors()).round()*2-1
	labels=Labels(array(lab))

	lda=LDA(gamma, feats_train, labels)
	lda.parallel.set_num_threads(num_threads)
	lda.train()

	lda.get_bias()
	lda.get_w()
	lda.set_features(feats_test)
	lda.classify().get_labels()

###########################################################################
# call functions
###########################################################################

if __name__=='__main__':
	seed(42)

	svm_light()
	libsvm()
	gpbtsvm()
	mpdsvm()
	libsvm_multiclass()
	libsvm_oneclass()
	gmnpsvm()

	do_batch_linadd()

	subgradient_svm()
	svmocas()
	liblinear()
	svmlin()

	perceptron()
	knn()
	lda()

File: matlab_and_octave/examples/all_classifier.m
% Explicit examples on how to use the different classifiers

num=24;
dist=2.2;
len=42;
size_cache=10;
C=0.017;
use_bias=0;
epsilon=1e-5;
width=2.1;
max_train_time=60;

trainlab_one=[ones(1,num*2) -ones(1,num*2)];
trainlab_multi=[zeros(1,num) ones(1,num) 2*ones(1,num) 3*ones(1,num)];

traindata_real=[randn(2,num)-dist, randn(2,num)+dist, randn(2,num)+dist*[ones(1,num); zeros(1,num)], randn(2,num)+dist*[zeros(1,num); ones(1,num)]];
testdata_real=[randn(2,num+7)-dist, randn(2,num+7)+dist, randn(2,num+7)+dist*[ones(1,num+7); zeros(1,num+7)], randn(2,num+7)+dist*[zeros(1,num+7); ones(1,num+7)]];

acgt='ACGT';
trainlab_dna=[ones(1,num/2) -ones(1,num/2)];
traindata_dna=acgt(ceil(4*rand(len,num)));
testdata_dna=acgt(ceil(4*rand(len,num)));


%
% kernel-based SVMs
%

% SVMLight
disp('SVMLight');

degree=20;
sg('set_features', 'TRAIN', traindata_dna, 'DNA');
sg('send_command', sprintf('set_kernel WEIGHTEDDEGREE CHAR %d %d', size_cache, degree));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_dna);
sg('send_command', 'new_svm LIGHT');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_dna, 'DNA');
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% LibSVM
disp('LibSVM');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm LIBSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% GPBTSVM
disp('GPBTSVM');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm GPBTSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% MPDSVM
disp('MPDSVM');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm MPDSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% LibSVM MultiClass
disp('LibSVMMultiClass');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_multi);
sg('send_command', 'new_svm LIBSVM_MULTICLASS');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% LibSVM OneClass
disp('LibSVMOneClass');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('send_command', 'new_svm LIBSVM_ONECLASS');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% GMNPSVM
disp('GMNPSVM');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm GMNPSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');
result=sg('svm_classify');


% run with batch or linadd on LibSVM;
disp('LibSVM batch');

sg('set_features', 'TRAIN', traindata_real);
sg('send_command', sprintf('set_kernel GAUSSIAN REAL %d %f', size_cache, width));
sg('send_command', 'init_kernel TRAIN');

sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm LIBSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_kernel TEST');

objective=sg('get_svm_objective');
sg('send_command', 'use_batch_computation 1');
sg('send_command', 'use_linadd 1');
result=sg('svm_classify');


%
% SparseLinear classifier
%

% SubgradientSVM - often does not converge
disp('SubGradientSVM');

C=0.42;
sg('set_features', 'TRAIN', sparse(traindata_real));
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm SUBGRADIENTSVM');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', sprintf('svm_max_train_time %d', max_train_time));
% sometimes does not terminate
%sg('send_command', 'svm_train');

%sg('set_features', 'TEST', sparse(testdata_real));
%result=sg('svm_classify');

% SVMOcas
disp('SVMOcas');

sg('set_features', 'TRAIN', sparse(traindata_real));
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm SVMOCAS');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', sparse(testdata_real));
result=sg('svm_classify');

% SVMSGD
disp('SVMSGD');

sg('set_features', 'TRAIN', sparse(traindata_real));
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm SVMSGD');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', sprintf('svm_max_train_time %d', max_train_time));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', sparse(testdata_real));
result=sg('svm_classify');

% LibLinear
disp('LibLinear');

sg('set_features', 'TRAIN', sparse(traindata_real));
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm LIBLINEAR_LR');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', sprintf('svm_max_train_time %d', max_train_time));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', sparse(testdata_real));
result=sg('svm_classify');

% SVMLin
disp('SVMLin');

sg('set_features', 'TRAIN', sparse(traindata_real));
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_svm SVMLIN');
sg('send_command', sprintf('svm_epsilon %f', epsilon));
sg('send_command', sprintf('c %f', C));
sg('send_command', sprintf('svm_use_bias %d', use_bias));
sg('send_command', sprintf('svm_max_train_time %d', max_train_time));
sg('send_command', 'svm_train');

sg('set_features', 'TEST', sparse(testdata_real));
result=sg('svm_classify');


%
% misc classifiers
%

% Perceptron
disp('Perceptron');

sg('set_features', 'TRAIN', traindata_real);
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_classifier PERCEPTRON');
sg('send_command', 'train_classifier');

sg('set_features', 'TEST', testdata_real);
result=sg('classify');

% KNN
disp('KNN');

sg('set_features', 'TRAIN', traindata_real);
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'set_distance EUCLIDIAN REAL');
sg('send_command', 'init_distance TRAIN');
sg('send_command', 'new_knn');
sg('send_command', 'train_knn');

sg('set_features', 'TEST', testdata_real);
sg('send_command', 'init_distance TEST');
result=sg('classify');


% LDA
disp('LDA');

sg('set_features', 'TRAIN', traindata_real);
sg('set_labels', 'TRAIN', trainlab_one);
sg('send_command', 'new_classifier LDA');
sg('send_command', 'train_classifier');

sg('set_features', 'TEST', testdata_real);
result=sg('classify');

